diff --git a/ffreed/args.py b/ffreed/args.py
index 7d29847..cfdac6d 100644
--- a/ffreed/args.py
+++ b/ffreed/args.py
@@ -9,10 +9,8 @@ from ffreed.utils import lmap
 def str2strs(s):
     return s.split(',')
 
-
 def str2floats(s):
-    return lmap(float, s.split(','))
-
+    return lmap(float, s.split(' '))
     
 def str2ints(s):
     return lmap(int, s.split(','))
@@ -31,7 +29,7 @@ def parse_args():
     parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
 
     parser.add_argument('--name', type=str, required=True)
-    parser.add_argument('--exp_root', type=str, default='/mnt/2tb/experiments/freed')
+    parser.add_argument('--exp_root', type=str, default='../experiments')
     parser.add_argument('--commands', type=str2strs, default='train,sample')
     parser.add_argument('--device', type=str, default='cuda:0')
     parser.add_argument('--local_rank', type=int, default=0)
@@ -42,6 +40,7 @@ def parse_args():
     parser.add_argument('--fragmentation', type=str, default='crem', choices=['crem', 'brics'])
     parser.add_argument('--gamma', type=float, default=0.99)
     parser.add_argument('--starting_smile', type=str, default='c1([*:1])c([*:2])ccc([*:3])c1')
+    parser.add_argument('--scaffolds_json', type=str)
     parser.add_argument('--timelimit', type=int, default=4)
 
     # model updating
@@ -79,7 +78,7 @@ def parse_args():
 
     # Docking
     parser.add_argument('--receptor', required=True)
-    parser.add_argument('--box_center', required=True, type=str2floats)
+    parser.add_argument('--box_center', required=True, type=str2floats, nargs='*')
     parser.add_argument('--box_size', required=True, type=str2floats)
     parser.add_argument('--vina_program', required=True)
     parser.add_argument('--exhaustiveness', type=int, default=8)
diff --git a/ffreed/env/reward.py b/ffreed/env/reward.py
index 5447c53..4d69546 100644
--- a/ffreed/env/reward.py
+++ b/ffreed/env/reward.py
@@ -1,6 +1,10 @@
 from functools import partial
+import pickle
+import numpy as np
 from rdkit import Chem
+from rdkit.Chem import MACCSkeys
 from rdkit.Chem import rdMolDescriptors
+from joblib import load
 
 
 class Reward:
@@ -55,3 +59,56 @@ class PatternFilter:
 
 def MolLogP(m):
     return rdMolDescriptors.CalcCrippenDescriptors(m)[0]
+
+def smi_to_maccs(smiles):
+    mols = [Chem.MolFromSmiles(smile) for smile in smiles]
+    fps = []
+    for i in range(len(mols)):
+        fp = [int(bit) for bit in MACCSkeys.GenMACCSKeys(mols[i]).ToBitString()][:167]
+        fps.append(fp)
+    fps_array = np.array(fps, dtype=np.int32).reshape(1, -1) 
+    return fps_array
+
+def smi_to_descriptors(smiles):
+    descriptor_names = list(rdMolDescriptors.Properties.GetAvailableProperties())
+    get_descriptors = rdMolDescriptors.Properties(descriptor_names)
+    
+    mols = [Chem.MolFromSmiles(smile) for smile in smiles]
+    descriptors = []
+    
+    for mol in mols:
+        if mol:  # If molecule is valid
+            desc = np.array(get_descriptors.ComputeProperties(mol))
+        else:  # If molecule is invalid
+            desc = np.zeros(len(descriptor_names))  # Zero-filled array for invalid molecules
+        descriptors.append(desc)
+    descriptors_array = np.array(descriptors, dtype=np.float32)
+
+    print("Descriptor names:", descriptor_names)
+    
+    return descriptors_array
+
+
+def Melanin(mol):
+    model = load("melanin.pkl")
+    fps = smi_to_maccs([Chem.MolToSmiles(mol)])
+    mel = model.predict_proba(fps)[0][1]
+    return mel
+
+
+def Irritation(mol):
+    model = pickle.load(open('irritation.pkl', 'rb'))
+    fps = smi_to_maccs([Chem.MolToSmiles(mol)])
+    irr = model.predict_proba(fps)[0][1]
+    return irr
+
+
+def Corneal(mol): 
+    model = load("corneal.pkl")
+    desc = smi_to_descriptors([Chem.MolToSmiles(mol)]) 
+    print("Descriptor array shape:", desc.shape)
+    cor = model.predict_proba(desc)[0][1]
+    return cor
+
+
+
diff --git a/ffreed/main.py b/ffreed/main.py
index 4974a06..9b3d1e4 100755
--- a/ffreed/main.py
+++ b/ffreed/main.py
@@ -17,7 +17,8 @@ from rdkit.Chem import Lipinski, MolFromSmiles
 from rdkit import RDLogger
 RDLogger.DisableLog('rdApp.*')
 
-from ffreed.env.reward import Reward, MolLogP, PatternFilter, OutOfRange, identity
+from ffreed.env.reward import Reward, MolLogP, PatternFilter, OutOfRange, identity, Corneal, Melanin, Irritation
+
 from ffreed.env.docking import DockingVina
 from ffreed.env.environment import Environment
 from ffreed.train.sac import SAC
@@ -60,6 +61,9 @@ def init_rewards(args):
         'PAINS': Reward(PatternFilter(patterns['PAINS']), identity, preprocess=preprocess),
         'SureChEMBL': Reward(PatternFilter(patterns['SureChEMBL']), identity, preprocess=preprocess),
         'Glaxo': Reward(PatternFilter(patterns['Glaxo']), identity, preprocess=preprocess),
+        'Corneal': Reward(Corneal, OutOfRange(lower=0.9, upper=1.0, hard=hard), preprocess=preprocess),
+        'Melanin': Reward(Melanin, OutOfRange(lower=0.9, upper=1.0, hard=hard), preprocess=preprocess),
+        'Irritation': Reward(Irritation, OutOfRange(lower=0.0, upper=0.1, hard=hard), preprocess=preprocess),
     }
     rewards = dict()
     for name, weight in zip(args['objectives'], args['weights']):
